text/0002-observables.md

```diff
@@ -0,0 +1,472 @@
+
+
+#### Redux
+
+```js
+// src/redux.js
+import { createStore } from 'redux';
+
+export const store = createStore((state = 0, action) => {
+  switch (action.type) {
+    case 'INCREMENT':
+      return state + 1
+    case 'DECREMENT':
+      return state - 1
+    default:
+      return state
+  }
+});
+
+function adaptor(store) {
+  return {
+    subscribe(fn) {
+      return store.subscribe(() => {
+        fn(store.getState());
+      });
+    }
+  };
+}
+
+export const observable = adaptor(store);
+```
+
+```html
+<!-- src/Counter.html -->
+<script>
+  import { store, observable } from './redux.js';
+</script>
+
+<button on:click="{() => store.dispatch({ type: 'INCREMENT' })}">
+  Clicks: {$observable}
+</button>
+```
+
+
+#### Immer
+
+```js
+import { produce } from 'immer';
+
+function immerObservable(data) {
+  const o = observable(data);
+
+  function update(fn) {
+    o.update(state => produce(state, fn));
+  }
+
+  return {
+    update,
+    subscribe: o.subscribe
+  };
+}
+
+const todos = immerObservable([
+  { done: false, description: 'walk the dog' },
+  { done: false, description: 'mow the lawn' },
+  { done: false, description: 'do the laundry' }
+]);
+
+todos.update(draft => {
+  draft[0].done = true;
+});
+```
+
+
+#### Shiz
+
+```js
+import { value, computed } from 'shiz';
+
+const a = value(1);
+const b = computed([a], ([a]) => a * 2);
+
+function shizObservable(shiz) {
+  return readOnlyObservable(function start(set) {
+    return shiz.on('change', () => {
+      set(shiz.get());
+    });
+  }, shiz.get());
+}
+
+const observable = shizObservable(b);
+
+const unsubscribe = observable.subscribe(value => {
+  console.log(value); // logs 2
+});
+
+a.set(2); // logs 4
+```
+
+
+### Using with Sapper
+
+At present, `Store` gets privileged treatment in [Sapper](https://sapper.svelte.technology) apps. A store instance can be created per-request, for example to contain user data. This store is attached to the component tree at render time, allowing `<span>{$user.name}</span>` to be server-rendered; its data is then passed to a client-side store.
+
+It's essential that this functionality be preserved. I'm not yet sure of the best way to achieve that.
+
+> Potential corner-cases to discuss:
+> * What happens if a subscriber causes another subscriber to be removed (e.g. it results in a component subtree being destroyed)?
+> * Is `$user.name` ambiguous (i.e. is `user` the observable, or `user.name`?) and if so how do we resolve the ambiguity
+> * What happens if `$user` is declared in a scope that has an observable `user`? Do we just not subscribe?
+
+
+
+### Where does it go?
+
+This is (🐃) up for debate, but one possibility is that we put these functions in `svelte/state`:
+
+```js
+import { observable, readOnlyObservable, derivedObservable } from 'svelte/state';
```

Comment:
```
import { observe, readonly, derived } from 'svelte/store';
import { observeData, readonlyData, derivedData } from 'svelte/store';
import { observeStore, readonlyStore, derivedStore } from 'svelte/store';
```
Sorry, but don't agree with this comment at all:
```
It's arguably simpler to teach than the existing store ...
```
this implementation is much more complicated imo but brings in readonly and a number of other benefits

Reply:
Yeah, I like these much better. I wonder about `derive` or `compute` rather than `merge`? But no strong opinions.

Reply:
>  I wonder about derive or compute rather than merge? 

`compose`?

Reply:
Yeah, I like that a lot.

Relatedly though, one thing I was thinking about is that this RFC doesn't yet show a good way of thinking about async values. One option is simply to model them as streams of promises:

```html
<script>
  import { source, compose } from 'svelte/store';

  const search = source('');

  const suggestions = compose(search, async q => {
    const r = await fetch(`api/suggestions?q=${encodeURIComponent(q)}`;
    return await r.json();
  });
</script>

<input
  list="suggestions"
  value={$search}
  on:input="{e => search.set(e.target.value}"
>

{#if $search}
  {#await $suggestions then suggestions}
    <datalist id="suggestions">
      {#each suggestions as suggestion}
        <option>{suggestion}</option>
      {/each}
    </datalist>
  {/await}
{/if}
```

But in some situations you might want to model it as a stream of non-promise values. In those cases, you'd want your composer to `set` new values, rather than returning them:

```js
function alternativeCompose(...args) {
  const fn = args.pop();

  return readonly(set => {
    let running = false;
    const values = [];

    const unsubscribers = args.map((arg, i) => arg.subscribe(value => {
      values[i] = value;
      if (running) fn(...values, set);
    }));

    running = true;
    fn(...values, set);

    return function stop() {
      running = false;
      unsubscribers.forEach(fn => fn());
    };
  });
}

const suggestions = alternativeCompose(search, async (search, set) => {
  const r = await fetch(`api/suggestions?q=${encodeURIComponent(q)}`;
  set(await r.json());
});
```

Of course, the existing `compose` can be expressed using `alternativeCompose`...

```js
function compose(...args) {
  const fn = args.pop();
  return alternativeCompose(...args, (...args2) => {
    const set = args2.pop();
    set(fn(...args2));
  });
}
```

...in other words it's more 'foundational'. So maybe `compose` should be reserved for that, and the simpler (to use) function should be something else like `map`? Though 'map' implies a single input... gah, naming things is hard.

(Not shown in the `alternativeCompose` implementation: race condition handling and a default initial value.)

Reply:
`combine`?

Reply:
I'd be confused by `compose` as it's a functional programming term to mean [composing functions](https://ramdajs.com/docs/#compose), while here it'd be used to merge the observables with the result of running the last arg, a function.

`merge` (https://ramdajs.com/docs/#merge, https://ascartabelli.github.io/lamb/module-lamb.html#merge) might well signify that we rewrite some of the values, so it'd be also a no-go for me.

I always really liked `compute`and I don't see why changing it ;) but, if you really must, `extend` would also be clear and unconfusing to me (somehow Lodash aliased it with [assignIn](https://lodash.com/docs#assignIn) but that is a merge too as their `merge` is a recursive `assignIn`).
In general it should communicate well that we're adding a dependendent prop.

Reply:
Re: using `svelte/state`: I'd prefer if we sticked to "store".

To me "state" represents the state of the application or of a component ("LOADING", "DRAGGING", "CHECKING_DATA"), rather than the content of a store.
In my experience the `state` of the application is stored in a state machine which I usually store as a value of the Store, along with the current data (that depends on the current state of the application).

Reply:
Check the comments towards the bottom of the main thread — we've been thinking about

```js
import { readable, writable, derive } from 'svelte/source';
```